Backwards compatible pure data model API

Gozala opened the pull request:
This pull request attempts to add pure Data Model API interop (as per #173) in a backwards compatible way. High level overview of the changes are:

- Getters / Setters are replaced by non-writable public properties.
- Some internal ops that add / remove links do mutate arrays in place instead of swapping arrays (due to introduce non-writable constraint).
- serialization code now handles ArrayBuffer and TypedArray for `node.Data`.

Only visible change is that `node.Links` no longer returns links mapped to pure data, but rather `DAGLink` instances. However it should not matter in practice because those were made to behave just like pure data would. Although tests had to be change to use `.containSubset` instead of `.eql` because later fails when prototype chain is different.

vmx commented:
Wow, nice work. I just tested it also with `js-ipld` and it just worked. If @achingbrain is happy with those changes, I don't see a reason for not making those changes.

achingbrain commented on test/dag-node-test.js:
```diff
@@ -4,8 +4,10 @@
 const chai = require('chai')
 const { Buffer } = require('buffer')
 const dirtyChai = require('dirty-chai')
+const chaiSubset = require('chai-subset')
```
You can require chai from `aegir/utils/chai` and it has `dirty-chai` and `chai-subset` preconfigured. It also avoids nasty weird bugs related to the order of `chai.use` invocations [across the entire codebase](https://github.com/chaijs/chai/issues/1298).

achingbrain commented on test/dag-node-test.js:
```diff
@@ -80,7 +82,7 @@ module.exports = (repo) => {
 
       const node2 = new DAGNode(someData, l2)
-      expect(node2.Links).to.eql([l1[1], l1[0]])
+      expect(node2.Links).to.containSubset([l1[1], l1[0]])
```
Doesn't marking the properties non-enumerable mean you should be able to use `.deep.equal` (or `.eql`) here?

Gozala replied:
I believe problem with `.deep.equal` and `.eql` is that fails if values do have different prototype chains.

Gozala replied:
Turns out the problem here was caused by `[Symbol.toStringTag]` getter added by `class-is` library.

https://github.com/moxystudio/js-class-is/blob/master/index.js#L19-L21

I'm assuming `eql` somewhere does `Object.prototype.toString.call(v)` but unless this is a blocker, I would prefer not to dig into this any further.

achingbrain replied:
Could you open an issue against `class-is` please?  It should not get in the way like this.

Gozala replied:
https://github.com/moxystudio/js-class-is/issues/27

achingbrain commented on src/serialize.js:
```diff
@@ -20,7 +20,7 @@ const toProtoBuf = (node) => {
     pbn.Links = node.Links
       .map((link) => ({
-        Hash: link.Hash.buffer,
+        Hash: asBuffer(link.Hash.buffer),
```
Did you mean to pass `link.Hash` in here instead?

If `link.Hash` is a node buffer, the `.buffer` property is an ArrayBuffer that might not correspond exactly to the original bytes.

We need to use `Buffer.from`, `link.Hash.byteOffset` and `link.Hash.length` to make sure we don't end up with the wrong bytes if we're accessing the backing ArrayBuffer directly - https://nodejs.org/api/buffer.html#buffer_buf_buffer

As it stands it'll get turned back into a node Buffer further down at line 44 that contains the whole ArrayBuffer.

Gozala replied:
Good catch, I meant `link.Hash` indeed (I think I've overlooked `link.Hash` was Buffer rather than thing that had `{buffer: Buffer}`).

_Fact that no test caught this makes me quite a bit nervous._

Gozala replied:
Wait no, `link.Hash` is a `CID` not the `Buffer`

Gozala replied:
New commit removes this changes in favor of adding Uint8Array support into protons library instead (See https://github.com/ipfs/protons/pull/13)

achingbrain replied:
> Wait no, link.Hash is a CID not the Buffer

Quite right, this is why I shouldn't look at these things in the evening!

achingbrain commented on src/dag-node/sortLinks.js:
```diff
@@ -8,12 +8,23 @@ const linkSort = (a, b) => {
  * @param {Array} links
  * @returns {Array}
  */
 const sortLinks = (links) => {
```
If we're switching to in-place sorting, is this function used somewhere else in the module or can it be removed?

Gozala replied:
Only reason I have not done it is I do not know if anything outside of the package might be making use of this. If we're comfortable removing it, I'm happy to do so

Gozala replied:
Made a change that makes replaces existing `sortLinks` with mutable sortLinks`. One thing that makes me bit uncomfortable about this change is that it's default import and change will in behavior will not be immediately obvious (if it was named, I'd just renamed the function).

Gozala commented:
Summary of few updates I've made:

1. Pull `chai` from `aegir`.
2. Remove `Uint8Array` to `Buffer` casting in favor of doing it in protons (See https://github.com/ipfs/protons/pull/13)
  - Current patch changes protons dep to point to my fork. Once it lands and published I'll update this to point to new verison.
  - `ArrayBuffer` as `Data` is no longer supported, it's just `Uint8Array` now. Mostly to keep things simple on `protons` side.
3. Replaced immutable `sortLinks` with mutable version. Immutable version was not used anywhere else in this package.

vmx commented on package.json:
```diff
@@ -69,13 +69,11 @@
     "multicodec": "^1.0.1",
     "multihashing-async": "~0.8.1",
-    "protons": "^1.0.2",
+    "protons": "git://github.com/gozala/protons#uint8array",
```
I planned to switch to https://github.com/mapbox/pbf if we ever switch to UInt8Array support. It might be worth a look now. IIRC that version just works automatically on any array-like thing, hence Node.js Buffers as well as UInt8Arrays.

Gozala replied:
That sounds like a good idea! However I would much rather keep the rabbit hole as shallow as possible. That is to say can we do that independently of this pull request ?

P.S.: I obviously don't suggest landing this with my fork in dependencies, just have it here until this https://github.com/dignifiedquire/borc/pull/48 makes it's way to npm.

rvagg commented on src/dag-node/rmLink.js:
```diff
@@ -4,12 +4,26 @@ const CID = require('cids')
   // It's a name
   if (typeof nameOrCid === 'string') {
-    dagNode._links = dagNode._links.filter((link) => link.Name !== nameOrCid)
+    predicate = link => link.Name === nameOrCid
```
minor nit: the style in this library is to always-paren arrow functions, it'd be great to keep it that way (a rule I'd love enforced by standard tbh, for readability, but for now it's on a per-library convention basis). There's another instance 2 lines below too.

rvagg commented on src/dag-node/dagNode.js:
```diff
@@ -25,24 +25,26 @@ class DAGNode {
     }
 
-    links = links.map((link) => {
+    links = links.map(link => {
```
another instance of missing parens

Gozala commented:
@achingbrain I believe I've addressed all your comments and comments from @rvagg.  @vmx I hope you agree with my arguments to decouple this from switch to https://github.com/mapbox/pbf, I do not know what errors may arise from that change & I do not want to block my primary work on that.

Please let me know if there is anything else to be done, or if this can land.

Thanks

Gozala commented:
@vmx any chance you could get to this sometime this week ? I can't make any more progress on  ipfs/js-ipfs#3081 without this.

vmx commented:
@Gozala any objections if I release this as 0.19? It's been quite a big change and I don't want to have regressions sneaked it.

Gozala commented:
@vmx that sounds good to me. Thanks
